File: java/save-command/src/test/java/org/babyfish/jimmer/example/save/common/AbstractMutationTest.java
package org.babyfish.jimmer.example.save.common;

import org.babyfish.jimmer.meta.ImmutableProp;
import org.babyfish.jimmer.sql.JSqlClient;
import org.babyfish.jimmer.sql.dialect.H2Dialect;
import org.babyfish.jimmer.sql.runtime.*;
import org.h2.Driver;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;
import java.util.function.BiFunction;

public abstract class AbstractMutationTest {

    private Connection connection;

    private JSqlClient sqlClient;

    private List<ExecutedStatement> executedStatements;

    @BeforeEach
    public void beforeTest() throws SQLException {
        connection = createConnection();
        createDatabase(connection);
        executedStatements = new ArrayList<>();
        JSqlClient.Builder builder = JSqlClient
                .newBuilder()
                .setDialect(new H2Dialect())
                .setTargetTransferable(true)
                .setExecutor(new RecordSqlExecutor())
                .setConnectionManager(ConnectionManager.singleConnectionManager(connection));
        customize(builder);
        sqlClient = builder.build();
    }

    @AfterEach
    public void afterTest() throws SQLException {
        sqlClient = null;
        Connection con = connection;
        if (con != null) {
            connection = null;
            con.close();
        }
    }

    protected JSqlClient sql() {
        return sqlClient;
    }

    protected void jdbc(String sql, Object ... args) {
        try (PreparedStatement statement = connection.prepareStatement(sql)) {
            for (int i = 0; i < args.length; i++) {
                statement.setObject(i + 1, args[i]);
            }
            statement.execute();
        } catch (SQLException ex) {
            throw new IllegalArgumentException(ex);
        }
    }

    protected void assertExecutedStatements(ExecutedStatement... executedStatements) {
        int count = Math.min(this.executedStatements.size(), executedStatements.length);
        for (int i = 0; i < count; i++) {
            Assertions.assertEquals(
                    executedStatements[i].getSql(),
                    this.executedStatements.get(i).getSql(),
                    "Failed to assert sql of statements[" + i + "]"
            );
            Assertions.assertEquals(
                    executedStatements[i].getVariableLists().size(),
                    this.executedStatements.get(i).getVariableLists().size(),
                    "Failed to assert batch row count of statements[" + i + "]"
            );
            Assertions.assertEquals(
                    executedStatements[i].getSql(),
                    this.executedStatements.get(i).getSql(),
                    "Failed to assert sql of statements[" + i + "]"
            );
            for (int ii = 0; ii < executedStatements[i].getVariableLists().size(); ii++) {
                Assertions.assertEquals(
                        executedStatements[i].getVariableLists().get(ii),
                        this.executedStatements.get(i).getVariableLists().get(ii),
                        "Failed to assert variables of statements[" + i + "][" + ii + "]"
                );
            }
        }
        Assertions.assertEquals(
                executedStatements.length,
                this.executedStatements.size(),
                "Expected " +
                        executedStatements.length +
                        " statements, but " +
                        this.executedStatements.size() +
                        " statements"
        );
    }

    // Can be overridden
    protected void customize(JSqlClient.Builder builder) {}

    private static Connection createConnection() throws SQLException {
        Properties properties = new Properties();
        properties.setProperty("database_to_upper", "true");
        return new Driver().connect(
                "jdbc:h2:mem:save-cmd-example",
                properties
        );
    }

    private void createDatabase(Connection con) {
        InputStream stream = AbstractMutationTest.class.getClassLoader().getResourceAsStream("unit-test.sql");
        if (stream == null) {
            Assertions.fail("Failed to initialize database, cannot load 'database.sql'");
        }
        try (Reader reader = new InputStreamReader(stream)) {
            StringBuilder builder = new StringBuilder();
            char[] buf = new char[1024];
            int len;
            while ((len = reader.read(buf)) != -1) {
                builder.append(buf, 0, len);
            }
            con.createStatement().execute(builder.toString());
        } catch (IOException | SQLException ex) {
            Assertions.fail("Failed to initialize database", ex);
        }
    }

    private class RecordSqlExecutor implements Executor {

        @Override
        public <R> R execute(Args<R> args) {
            executedStatements.add(
                    ExecutedStatement.of(
                            args.sql,
                            args.variables.toArray(new Object[0])
                    )
            );
            return DefaultExecutor.INSTANCE.execute(args);
        }

        public Executor.BatchContext executeBatch(
                Connection con,
                String sql,
                @Nullable ImmutableProp generatedIdProp,
                @NotNull ExecutionPurpose purpose,
                @NotNull JSqlClientImplementor sqlClient
        ) {
            BatchContext raw = DefaultExecutor.INSTANCE.executeBatch(
                    con,
                    sql,
                    generatedIdProp,
                    purpose,
                    sqlClient
            );
            return new BatchContextWrapper(raw, sql, executedStatements);
        }
    }

    private static class BatchContextWrapper implements Executor.BatchContext {

        private final Executor.BatchContext raw;

        private final String sql;

        private final List<ExecutedStatement> executedStatements;

        private final List<List<Object>> variableLists = new ArrayList<>();

        BatchContextWrapper(
                Executor.BatchContext raw,
                String sql,
                List<ExecutedStatement> executedStatements
        ) {
            this.raw = raw;
            this.sql = sql;
            this.executedStatements = executedStatements;
        }

        @Override
        public JSqlClientImplementor sqlClient() {
            return raw.sqlClient();
        }

        @Override
        public String sql() {
            return raw.sql();
        }

        @Override
        public ExecutionPurpose purpose() {
            return raw.purpose();
        }

        @Override
        public ExecutorContext ctx() {
            return raw.ctx();
        }

        @Override
        public void add(List<Object> variables) {
            variableLists.add(variables);
            raw.add(variables);
        }

        @SuppressWarnings("unchecked")
        @Override
        public int[] execute(
                BiFunction<SQLException, Executor.BatchContext, Exception> exceptionTranslator
        ) {
            executedStatements.add(
                    ExecutedStatement.batchOf(
                            sql,
                            variableLists.toArray((List<Object>[])new List[0])
                    )
            );
            return raw.execute(exceptionTranslator);
        }

        @Override
        public Object[] generatedIds() {
            return raw.generatedIds();
        }

        @Override
        public void addExecutedListener(Runnable listener) {
            raw.addExecutedListener(listener);
        }

        @Override
        public void close() {
            raw.close();
        }
    }
}

File: kotlin/save-command-kt/src/test/kotlin/org/babyfish/jimmer/example/save/common/AbstractMutationTest.kt
package org.babyfish.jimmer.example.save.common

import org.babyfish.jimmer.meta.ImmutableProp
import org.babyfish.jimmer.sql.dialect.H2Dialect
import org.babyfish.jimmer.sql.kt.KSqlClient
import org.babyfish.jimmer.sql.kt.cfg.KSqlClientDsl
import org.babyfish.jimmer.sql.kt.newKSqlClient
import org.babyfish.jimmer.sql.runtime.DefaultExecutor
import org.babyfish.jimmer.sql.runtime.ExecutionPurpose
import org.babyfish.jimmer.sql.runtime.Executor
import org.babyfish.jimmer.sql.runtime.Executor.BatchContext
import org.babyfish.jimmer.sql.runtime.JSqlClientImplementor
import org.h2.Driver
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import java.io.IOException
import java.io.InputStreamReader
import java.lang.Exception
import java.sql.Connection
import java.sql.SQLException
import java.util.*
import java.util.function.BiFunction
import kotlin.math.min

abstract class AbstractMutationTest {

    private lateinit var connection: Connection

    private lateinit var sqlClient: KSqlClient

    private lateinit var executedStatements: MutableList<ExecutedStatement>

    @BeforeEach
    @Throws(SQLException::class)
    fun beforeTest() {
        connection = createConnection()
        createDatabase(connection)
        executedStatements = mutableListOf()
        sqlClient = newKSqlClient {
            setDialect(H2Dialect())
            setConnectionManager {
                proceed(connection)
            }
            setTargetTransferable(true)
            setExecutor(
                object : Executor {
                    override fun <R : Any?> execute(args: Executor.Args<R>): R {
                        executedStatements.add(
                            ExecutedStatement.of(args.sql, *args.variables.toTypedArray())
                        )
                        return DefaultExecutor.INSTANCE.execute(args)
                    }

                    override fun executeBatch(
                        con: Connection,
                        sql: String,
                        generatedIdProp: ImmutableProp?,
                        purpose: ExecutionPurpose,
                        sqlClient: JSqlClientImplementor
                    ): BatchContext {
                        val variableLists = mutableListOf<List<Any>>()
                        val raw = DefaultExecutor.INSTANCE.executeBatch(
                            con,
                            sql,
                            generatedIdProp,
                            purpose,
                            sqlClient
                        )
                        return object : BatchContext by (raw) {
                            override fun add(variables: List<Any>) {
                                variableLists.add(variables)
                                raw.add(variables)
                            }

                            override fun execute(
                                exceptionTranslator: BiFunction<SQLException, BatchContext, Exception>?
                            ): IntArray {
                                executedStatements.add(ExecutedStatement.batchOf(sql, *variableLists.toTypedArray()))
                                return raw.execute(exceptionTranslator)
                            }
                        }
                    }
                }
            )
            customize(this)
        }
    }

    @AfterEach
    @Throws(SQLException::class)
    fun afterTest() {
        connection.close()
    }

    protected val sql: KSqlClient
        get() = sqlClient

    protected fun jdbc(sql: String, vararg args: Any?) {
        connection.prepareStatement(sql).use { statement ->
            for (i in args.indices) {
                statement.setObject(i + 1, args[i])
            }
            statement.execute()
        }
    }

    protected fun assertExecutedStatements(vararg executedStatements: ExecutedStatement) {
        val count = min(this.executedStatements.size, executedStatements.size)
        for (i in 0 until count) {
            Assertions.assertEquals(
                executedStatements[i].sql,
                this.executedStatements[i].sql,
                "Failed to assert sql of statements[$i]"
            )
            Assertions.assertEquals(
                executedStatements[i].variableLists.size,
                this.executedStatements[i].variableLists.size,
                "Failed to assert batch row count of statements[$i]"
            )
            for (ii in 0 until executedStatements[i].variableLists.size) {
                Assertions.assertEquals(
                    executedStatements[i].variableLists[ii],
                    this.executedStatements[i].variableLists[ii],
                    "Failed to assert variables of statements[$i][$ii]"
                )
            }
        }
        Assertions.assertEquals(
            executedStatements.size,
            this.executedStatements.size,
            "Expected " +
                executedStatements.size +
                " statements, but " +
                this.executedStatements.size +
                " statements"
        )
    }

    protected open fun customize(dsl: KSqlClientDsl) {}

    companion object {

        private fun createConnection(): Connection {
            val properties = Properties()
            properties.setProperty("database_to_upper", "true")
            return Driver().connect(
                "jdbc:h2:mem:save-cmd-example",
                properties
            )
        }

        private fun createDatabase(con: Connection) {
            val stream = AbstractMutationTest::class.java.classLoader.getResourceAsStream("unit-test.sql")
            if (stream == null) {
                Assertions.fail<Unit>("Failed to initialize database, cannot load 'database.sql'")
            }
            try {
                InputStreamReader(stream).use { reader ->
                    val builder: StringBuilder = StringBuilder()
                    val buf: CharArray = CharArray(1024)
                    var len: Int
                    while ((reader.read(buf).also { len = it }) != -1) {
                        builder.append(buf, 0, len)
                    }
                    con.createStatement().execute(builder.toString())
                }
            } catch (ex: IOException) {
                Assertions.fail("Failed to initialize database", ex)
            } catch (ex: SQLException) {
                Assertions.fail("Failed to initialize database", ex)
            }
        }

        @JvmStatic
        protected fun String.toOneLine(): String =
            replace("\n", "")
                .replace("\r", "")
    }
}